lightning/src/ln/msgs.rs

```diff
@@ -918,24 +1007,60 @@ impl_writeable!(FundingLocked, 32+33, {
 	next_per_commitment_point
 });
 
-impl_writeable_len_match!(GlobalFeatures, {
-		{ GlobalFeatures { ref flags }, flags.len() + 2 }
-	}, {
-	flags
-});
+impl<T: FeatureContext> Writeable for Features<T> {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		w.size_hint(self.flags.len() + 2);
+		(self.flags.len() as u16).write(w)?;
+		for f in self.flags.iter().rev() { // We have to swap the endianness back to BE for writing
+			f.write(w)?;
+		}
+		Ok(())
+	}
+}
 
-impl_writeable_len_match!(LocalFeatures, {
-		{ LocalFeatures { ref flags }, flags.len() + 2 }
-	}, {
-	flags
-});
+impl<R: ::std::io::Read, T: FeatureContext> Readable<R> for Features<T> {
+	fn read(r: &mut R) -> Result<Self, DecodeError> {
+		let mut flags: Vec<u8> = Readable::read(r)?;
+		flags.reverse(); // Swap to big-endian
+		Ok(Self {
+			flags,
+			mark: PhantomData,
+		})
+	}
+}
 
-impl_writeable_len_match!(Init, {
-		{ Init { ref global_features, ref local_features }, global_features.flags.len() + local_features.flags.len() + 4 }
-	}, {
-	global_features,
-	local_features
-});
+impl Writeable for Init {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		// global_features gets the bottom 13 bits of our features, and local_features gets all of
+		// our relevant feature bits. This keeps us compatible with old nodes.
+		let global_len = cmp::min(2, self.features.flags.len());
+		w.size_hint(self.features.flags.len() + global_len + 4);
+		(global_len as u16).write(w)?;
+		for i in (0..global_len).rev() {
+			if i == 0 {
+				self.features.flags[i].write(w)?;
+			} else {
+				(self.features.flags[i] & ((1 << (14 - 8)) - 1)).write(w)?;
+			}
+		}
+		self.features.write(w)
```

Comment:
Not sure if this respect the spec, encoding the whole logical features instead of two bitmaps ? Legacy global features bits are going to be encoded twice, which seems contrary to "SHOULD use the minimum length required to represent the features field."

Reply:
Hmm. The way I read that is just that you SHOULDn't pad with 0 bytes, not that you should change the meaning of the data you send. I think the spec is pretty mum on the issue here - it doesn't say which bits you can set in globablfeatures, just that you cant send bits higher than 13. In practice, I think this should be fine - if there's some unknown bit the peer will do the right thing and ignore it if its odd or disconnect if its even.